documentation/specs/src/base-ledger/replay-protection.md

```diff
@@ -0,0 +1,100 @@
+    /// Transaction counter for replay-protection
+    pub tx_counter: u32,
+    /// the encrypted payload
+    pub inner_tx: EncryptedTx,
+    /// sha-2 hash of the inner transaction acting as a commitment
+    /// the contents of the encrypted payload
+    pub tx_hash: Hash,
+}
+``` 
+
+In addition, we need a counter in the storage subspace of every implict address:
+
+```
+/$Address/tx_counter: u32
+```
+
+In `process_proposal` we check that `tx_counter` field in `WrapperTx` is greater or equal to the value currently in storage; if this doesn't hold, the transaction is considered invalid.
+
+At last, in `finalize_block`, the protocol updates the counter key in storage, increasing its value by one. Now, if a malicious user tried to replay this transaction, the `tx_counter` in the struct would no longer be greater or equal to the one in storage and the transaction would be deemed invalid. 
+
+Note that the address whose counter is involved in this process is the one signing the transaction (the same as the `pk` field of the struct).
+
+## InnerTx
+
+The `EncryptedTx` incapsulated inside `WrapperTx` should be protected too. That's because, if it wasn't, an attacker could extract the inner tx, rewrap it, and replay it.\
+To simplify this check we will perform it entirely in Wasm: the check of the counter will be carryed out by the validity predicates while the actual writing of the counter in storage will be done by the transactions themselves. 
+
+To do so, the `SignedTxData` attached to the transaction will hold the current value of the counter in storage. The transaction will simply read the value from storage and increase its value by one. The VP will then check the validity of the signature and, if it's deemed valid, will proceed to checking if the pre value of the counter in storage was less or equal than the one contained in the transaction data and if the post value of the key in storage has been incremented by one.
+
+In the specific case of a transfer, since MASP already comes with replay protection as part of the Zcash design (see the [MASP specs](https://specs.namada.net/masp.html) and [Zcash protocol specs](https://zips.z.cash/protocol/protocol.pdf)), the counter in `SignedTxData` should be optional.
+
+An alternative implementation could place the protection for the inner tx in protocol, just like the wrapper one. In this case we would need to extend the `Tx` struct to hold an additional optional field (again, because of MASP) for the transaction counter and the address involved in replay protection, like so:
+
+```rust
+pub struct Tx {
+    pub code: Vec<u8>,
+    pub data: Option<Vec<u8>>,
+    pub timestamp: DateTimeUtc,
+    pub tx_counter: Option<(Address, u32)>
+}
+```
+
+The check would run in `process_proposal` and the update in `finalize_block`, just like for the wrapper transaction. The drawback of this implementation is that it implies the need for an hard fork in case of a modification of the replay protection mechanism.
```

Comment:
I think another issue is that from protocol's perspective, it's not clear who is the source of an inner transaction as that depends on what happens in the code and the validity predicates that get triggered from it should decide when the counter is required and when it's not. However, it's not decidable when `tx_counter` should be `Some` (although we could tie it the specific whitelisted txs set we'll have in Namada, but it wouldn't work with programmable txs in future), or who the address should be (this is however a blocker)